### PR TITLE
[Feature] New community-team and pool-team relations

### LINKED ISSUE #10302: ✨ Add Community-Team and Pool-Team relations
## ✨ Feature

Communities and Pools need a one-to-one relationship to a Team, which will allow permissions (via Roles) to be grouped by Community and by Pool.

## 🕵️ Details

QUESTION: Should we ensure every past and future Community or Pool has a linked Team, or simply create Teams as necessary when assigning roles which need them?
1. Create and attach a team for every community and pool already in the database. Create event listeners which create and attach a team any time a new Community or Pool is created. From then on, we can assume the team field is non-null.
2. Existing pools and communities are left with no matching team. We ensure the api and/or client can interpret that has having no team members. The mutation (or seeders) which add team members to a Pool or Community must check for team existence, and create it first if necessary.

ANSWER: Lets go with number 2. If we use a polymorphic relationship, the database can't guarantee every Teamable resource has a Team. It therefore seems safest to assume that it may not, and simply interpret that as having no team members. The only place Teams need to be created should ultimately be in the mutations which assign roles (or seeders and tests, of course).

QUESTION: 
Teams need a unique `name` field, used by Laratrust. When creating teams for Communities and Pools, how should the `name` field for each team be generated? 
ANSWER: just use the name of the related data model, Community.id or Pool.id.

QUESTION: should the foreign key go on the Team table, or the Community/Pool tables? 
 - Putting the foreign key on the Team table requires using [Laravel polymorphism](https://laravel.com/docs/11.x/eloquent-relationships#one-to-one-polymorphic-relations), which adds some extra complexity and 'magic'. However, it guarantees that the Team is related to at most one resource.
 - Putting the foreign key in the Community and Pool tables is simpler (no polymorphism) and allows us to guarantee, if we want, that every Community/Pool has a Team. However, its now technically possible for one Team in the database to be related to multiple resources.
ANSWER: let's go with a Polymorphic relation. That's the way the Laravel framework intends this kind of relationship to work, and we don't want to fight the framework.

## 🎨 Design File

Include a link to the design file (if it exists).

### 🧑‍🎨 Designer

Include the author of the linked design file's name (or even better, their `@githubUsername`) for future reference.

## 📸 Screenshot

Add a screenshot of the new feature (if possible).

## 🙋‍♀️ Proposed Implementation

The relationship should probably be implemented as a "teamable" polymorphic relationship on Team. This will ensure every team is associated with a single resource, if any.

This schema attempts to simplify things by abstracting away the Team model, with the intention of later removing it entirely. The Laratrust team model is still required on the backend, acting as a kind of pivot table between RoleAssignment and a Teamable resource (Pool or Community), but it doesn't need to be visible to the client.

```
# This version attempts to abstract away the Team model, removing one extra layer for the client (and devs) to think about.
# Later we deprecate the Team model, then remove it entirely.

interface HasRoleAssignments {
  roleAssignments: [RoleAssignment!]
}
  
type Team implements HasRoleAssignments {
  id: ID!
  name: String! # represents a unique key for a team
  displayName: LocalizedString @rename(attribute: "display_name")
  description: LocalizedString
  departments: [Department] @belongsToMany
  contactEmail: Email @rename(attribute: "contact_email")
  pools: [Pool] @hasMany @canResolved(ability: "view")
  roleAssignments: [RoleAssignment!]
    @hasMany
    @canRoot(ability: "viewTeamMembers")
}
  
type Community implements HasRoleAssignments {
  ...
  roleAssignments: [RoleAssignment!]
}
  
type Pool implements HasRoleAssignments {
  ...
  roleAssignments: [RoleAssignment!]
}
  
type Role {
  id: ID!
  name: String! # represents a unique key for a role
  displayName: LocalizedString
  description: LocalizedString
  isTeamBased: Boolean
  roleAssignments: [RoleAssignment!]
}
  
# This model represents the polymorphic role_user table and allows querying the team relationship
type RoleAssignment {
  id: ID!
  role: Role
  team: HasRoleAssignments
  user: UserPublicProfile
}
```

The RoleAssignment team relationship _may_ need to be an Accessor instead of an Eloquent Relationship. It might be implementable with some kind of polymorphic hasOneThrough relationship, but the fact that sometimes it returns the pivot model (the Team itself) complicates things.

## 🌎 Localization

(optional) Provide any new copy along with translations available.

## ✅ Acceptance Criteria

- [ ] Communities and Pools can be associated with a single team.
- [ ] RoleAssignment team field in schema returns Pool or Community if the team is related to that resource
- [ ] Relationships exist on Eloquent models which allow for getting the community/pool assocated with a team, or the team associated with a community/pool, or the RoleAssignments associated with a community/pool.
- [ ] It is possible to query for the Community/Pool associated with the appropriate RoleAssignment, or query for the RoleAssignments associated with a Community/Pool, provided one has the correct permissions (see #10349)

## 🛑 Blockers

Issues which must be completed before this one.

```[tasklist]
### Blocked By
- [ ] #10298 
- [ ] https://github.com/GCTC-NTGC/gc-digital-talent/issues/10349
```

### PR DESCRIPTION
**Pull request by JamesHuf**
🤖 Resolves #10302 
## 👋 Introduction
Adds a new polymorphic "Teamable" relation connecting Teams to Communities and Pools

## 🕵️ Details
- There is an existing one-to-many Pool-Team relation which is being left alone for now.
- I created a new "teamable" field on RoleAssignments for returning it's teams Pool or Community if it has one since modifying the team field was conflicting with the relation

## 🧪 Testing
1. Run the new migration and reseed. Confirm this works as expected.
2. Run some test queries and confirm the new fields work as expected.

```
query teams {
  teams {
    teamable {
      __typename
    }
    roleAssignments {
      teamable {
        __typename
      }
    }
  }
}

query pools {
  poolsPaginated {
    data {
      team {
        id
        teamable {
          __typename
        }
        roleAssignments {
          id
        }
      }
      newTeam {
        id
        teamable {
          __typename
        }
        roleAssignments {
          id
        }
      }
      roleAssignments {
        id
      }
    }
  }
}

query communities {
  communities {
    team {
      teamable {
        __typename
      }
      roleAssignments {
        id
      }
    }
    roleAssignments {
      id
    }
  }
}
```

### CODE REVIEW

File: api/graphql/schema.graphql [hunk 1]
```diff
@@ -242,6 +247,7 @@ type Pool {
     @hasMany
     @canRoot(ability: "viewAssessmentPlan")
   poolSkills(type: PoolSkillType @where): [PoolSkill] @hasMany
+  roleAssignments: [RoleAssignment!] @hasManyThrough
```
Comment (JamesHuf) [inline api/graphql/schema.graphql:250]: I feel like I should check the viewTeamMembers policy here but I'm not sure how to call that on the relevant team. Should I create a similar policy on Pool?

File: api/graphql/schema.graphql [hunk 2]
```diff
@@ -391,21 +399,26 @@ type Department {
...
   name: LocalizedString
   description: LocalizedString
+  team: Team @morphOne @canModel(ability: "viewAny")
+  roleAssignments: [RoleAssignment!]
+    @hasManyThrough
+    @canRoot(ability: "viewTeamMembers")
 }
 
-type Team {
+type Team implements HasRoleAssignments {
```
Comment (petertgiles) [inline api/graphql/schema.graphql:413]: I can't really get anything out of the queries.  As soon as I add roleAssignments I start getting unauthorized exceptions.  Is is that there are no teams seeded for the pools or communities yet?

**Review by petertgiles — COMMENTED** [review]
To be honest, I'm still having trouble wrapping my head around this.  :grimacing: I've requested that Tristan look at this, too.  I'll try my best to keep reviewing once you've had a change to respond to my comments.

File: api/app/Models/Pool.php [hunk 3]
```diff
@@ -166,6 +168,18 @@ public function team(): BelongsTo
         return $this->belongsTo(Team::class);
     }
 
+    public function newTeam(): MorphOne
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply api/app/Models/Pool.php:171]: I agree. I like "legacyTeam".

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
It looks pretty good. I agree with Peter's comments, and just have a few of my own. 
The seeding still needs some work, which will make this easier to test, but here's also my updated query for testing:

```
fragment RoleAssignmentsF on RoleAssignment {
  role {
    name
  }
  user {
    email
  }
  teamable {
    __typename
    ... on Community {
      key
    }
    ... on Pool {
      name {
        en
      }
    }
  }
}

query teams {
  teams {
    name
    roleAssignments {
      ...RoleAssignmentsF
    }
  }
}

query pools {
  poolsPaginated {
    data {
      name {
        en
      }
      roleAssignments {
        ...RoleAssignmentsF
      }
    }
  }
}

query communities {
  communities {
		key
    roleAssignments {
      ...RoleAssignmentsF
    }
  }
}
```

File: api/app/Models/Pool.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/app/Models/Pool.php]: Updated in [3cc8de6](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/3cc8de66b2c0e4ed81f50dc5a0066cf0f2280663).

File: api/graphql/schema.graphql [hunk 4]
```diff
@@ -194,10 +194,15 @@ type ScreeningQuestion {
...
-type Pool {
+interface HasRoleAssignments {
+  roleAssignments: [RoleAssignment!]
+}
+
+type Pool implements HasRoleAssignments {
   id: ID!
   owner: UserPublicProfile @belongsTo(relation: "user")
   team: Team @belongsTo @canModel(ability: "viewAny")
+  newTeam: Team @belongsTo @canModel(ability: "viewAny") # Newer polymorphic team relationship
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/graphql/schema.graphql]: Updated as part of [3cc8de6](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/3cc8de66b2c0e4ed81f50dc5a0066cf0f2280663).

File: api/graphql/schema.graphql [hunk 5]
```diff
@@ -391,21 +399,26 @@ type Department {
   teams: [Team] @belongsToMany @canModel(ability: "viewAny")
 }
 
-type Community {
+type Community implements HasRoleAssignments {
   id: ID!
   key: String!
   name: LocalizedString
   description: LocalizedString
+  team: Team @morphOne @canModel(ability: "viewAny")
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/graphql/schema.graphql]: Updated as part of [3cc8de6](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/3cc8de66b2c0e4ed81f50dc5a0066cf0f2280663).

File: api/app/Policies/TeamPolicy.php [hunk 6]
```diff
@@ -84,4 +84,20 @@ public function assignTeamMembers(User $user, Team $team)
...
+    /**
+     * Determine whether the user can view the teams Teamable
+     *
+     * @return \Illuminate\Auth\Access\Response|bool
+     */
+    public function viewTeamable(User $user, Team $team)
+    {
+        $team->loadMissing('teamable');
+
+        if (is_null($team->teamable)) {
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/app/Policies/TeamPolicy.php]: The `can` check below this returns false if the model is null. I don't think it makes sense preventing anyone from viewing the lack of a teamable so I allow it here. I tried to make this slightly clearer in [d5b4d79](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/d5b4d7954aeba2c0f47298a3bd1bb4639477f7a9).

File: api/database/factories/CommunityFactory.php [hunk 7]
```diff
@@ -37,4 +38,32 @@ public function definition()
...
     //     });
     // }
+
+    /**
+     * Create a new community or get an existing community based on the given attributes.
+     *
+     * @param  array  $attributes  The attributes of the community.
+     * @return Community The created or existing community.
+     */
+    public function createOrGetExisting($attributes = [])
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/database/factories/CommunityFactory.php]: Updated this as part of [02a6879](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/02a68794c8095f76636abfc1e1529b6102f04f64).

File: api/graphql/schema.graphql (hunk 2 above)
Comment (JamesHuf) [reply api/graphql/schema.graphql:416]: Are you still seeing this with the new seeding?

File: api/database/factories/TeamFactory.php [hunk 8]
```diff
@@ -40,4 +41,11 @@ public function configure()
             $team->departments()->saveMany($departments);
         });
     }
+
+    public function withTeamMembers($count = 1)
+    {
+        return $this->afterCreating(function (Team $team) use ($count) {
+            User::factory()->asPoolOperator($team->name)->count($count)->create();
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/database/factories/TeamFactory.php]: Updated as part of [02a6879](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/02a68794c8095f76636abfc1e1529b6102f04f64).

File: api/database/seeders/CommunitySeeder.php [hunk 9]
```diff
@@ -32,14 +32,14 @@ public function run()
...
-                [
-                    'key' => $community['key'],
-                ],
-                [
-                    'name' => $community['name'],
-                ]
-            );
+            Community::factory()
+                ->withTeamMembers()
+                ->createOrGetExisting(
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/database/seeders/CommunitySeeder.php]: Updated as part of [02a6879](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/02a68794c8095f76636abfc1e1529b6102f04f64).

File: api/database/seeders/PoolTestSeeder.php [hunk 10]
```diff
@@ -30,6 +30,7 @@ public function run()
         $createdPool = Pool::factory()
             ->withPoolSkills(4, 4)
             ->withQuestions(2, 2)
+            ->withTeamMembers()
```
[the remark this answers is not in the document]
Comment (JamesHuf) [reply api/database/seeders/PoolTestSeeder.php]: Updated as part of [02a6879](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/02a68794c8095f76636abfc1e1529b6102f04f64).

File: api/app/Models/Community.php [hunk 11]
```diff
@@ -63,4 +63,54 @@ public function roleAssignments(): HasManyThrough
...
+    public function addCommunityRecruiter(string|array $userId)
+    {
+        $team = $this->team()->firstOrCreate([], [
+            'name' => "community-".$this->id,
+        ]);
+
+        if (is_array($userId)) {
+            foreach ($userId as $singleUserId) {
+                $user = User::find($singleUserId);
+                // $user->addRole('community_recruiter', $team->name);
```
Comment (tristan-orourke) [inline api/app/Models/Community.php:82]: If you were waiting for https://github.com/GCTC-NTGC/gc-digital-talent/pull/10533 for the new roles to be available, its not merged into main.

File: api/database/seeders/UserTestSeeder.php [hunk 12]
```diff
@@ -98,5 +102,40 @@ public function run()
                 'email' => 'noroles@test.com',
                 'sub' => 'noroles@test.com',
             ]);
+
+        User::factory()
+            ->asApplicant()
+            ->withSkillsAndExperiences()
```
Comment (tristan-orourke) [inline api/database/seeders/UserTestSeeder.php:108]: This is missing asProcessOperator. Also, I don't think this (or the community role users) need withSkillsAndExperiences.

File: api/database/factories/PoolFactory.php [hunk 13]
```diff
@@ -321,12 +321,23 @@ public function withBookmark(string $user_id)
...
-    // Add a team with at least one member to the pool
-    public function withTeamMembers()
+    /**
+     * Attach the users to the related pool.
+     * Creates a new user if no userIds passed in.
+     *
+     * @param array|null  $userIds - Id of the users to attach the role to
+     * @return void
+     */
+    public function withProcessOperators(array|null $userIds = null)
```
Comment (tristan-orourke) [inline api/database/factories/PoolFactory.php]: I like how this is written. Maybe add similar methods to CommunityFactory?

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Great job, this looks much better. Just need to use the proper roles in the `add<Role>` methods now.

File: api/app/Models/Community.php (hunk 11 above)
Comment (JamesHuf) [reply api/app/Models/Community.php]: I've updated to use the roles in [8f3f742](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/8f3f742c961de9a210e701b6d07fb8086f199877).

File: api/database/seeders/UserTestSeeder.php (hunk 12 above)
Comment (JamesHuf) [reply api/database/seeders/UserTestSeeder.php]: The process operator role is added in the PoolTestSeeder as the pool (and it's related team) don't exist yet. If you'd prefer I could probably update things to use a team defined in the TeamSeeder instead.

Removed withSkillsAndExperiences from the new users in [8739d66](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/8739d66b94b5afb5c37fe571d9e5e8167a78db6a).

File: api/database/factories/PoolFactory.php (hunk 13 above)
Comment (JamesHuf) [reply api/database/factories/PoolFactory.php]: Thanks! Added community variants in [9c8d27d](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/9c8d27d4c5e0b85ca2c0a347c9cedb50ee61fc85).

File: api/database/seeders/UserTestSeeder.php (hunk 12 above)
Comment (tristan-orourke) [reply api/database/seeders/UserTestSeeder.php]: Hmmm. Yes, I think you'd prefer to add it here. That would make it consistent with the Community-based roles and the PoolOperator role.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Very nice, just one comment left in UserTestSeeder

File: api/database/seeders/UserTestSeeder.php (hunk 12 above)
Comment (JamesHuf) [reply api/database/seeders/UserTestSeeder.php]: Changed seeding to attach the process operator role here in [533700e](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/533700e6cc4a3f20ab0b959ead9279da18604343).

File: api/database/factories/UserFactory.php [hunk 14]
```diff
@@ -294,20 +294,18 @@ public function asAdmin()
     /**
      * Attach the process operator role to a user after creation.
      *
-     * @param  string|array  $poolId  Id of the pool or pools to attach the role to
+     * @param  string|array  $team  Name of the team or teams to attach the role to
      * @return $this
      */
-    public function asProcessOperator(string|array $poolId)
+    public function asProcessOperator(string|array $team)
```
Comment (tristan-orourke) [inline api/database/factories/UserFactory.php:300]: Noooo, this was better before! 😭 

Just like `asCommunityRecruiter`, which takes in a communityId, this should take in a poolId.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Sorry, I feel like this last commit was a slight step backward. 

If UserTestSeeder runs before PoolTestSeeder, you could always create a single Pool in UserTestSeeder file using `Pool::factory()`, something basic, for the sole purpose of attaching to the Process Operator. Then, in PoolTestSeeder, you can still find the Process Operator user and attach them to more pools.

File: api/database/factories/UserFactory.php (hunk 14 above)
Comment (JamesHuf) [reply api/database/factories/UserFactory.php:300]: I prefer taking in a pool instead as well but the pool doesn't exist yet if I call this in UserTestSeeder.

Comment (JamesHuf) [reply api/database/factories/UserFactory.php]: How about [this](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10621/commits/1e2ad77ea2956fe8c2441aa152a086371aeaf755) instead?

**Review by tristan-orourke — APPROVED** [review]
Okay, great, I think this is ready to go now :)

**Pull request merged**